src/main/java/com/streamr/client/StreamrClient.java

```diff
@@ -190,81 +188,48 @@ public WebSocketClient getWebsocket() {
         return websocket;
     }
 
-    private void sleepThenReconnect() {
-        if (!stayConnected) { // TODO: remove?
-            return;
-        }
-        log.warn("Disconnected. Attempting to reconnect in " + options.getReconnectRetryInterval() / 1000 + " seconds.");
-        currentReconnectThread = new Thread(() -> {
-            try {
-                this.websocket.closeConnection(0, "");
-                Thread.sleep(options.getReconnectRetryInterval());
-                if (!stayConnected) { // TODO: remove?
-                    return;
-                }
-                this.reconnect();
-            } catch (InterruptedException e){
-                e.printStackTrace();
-                Thread.currentThread().interrupt();
-            }
-        }, "sleepThenReconnectThread-" + System.currentTimeMillis());
-        currentReconnectThread.start();
-    }
-
-    private void reconnect() {
-        initWebsocket();
-        websocket.connect();
-        waitForState(ReadyState.OPEN);
-        if (getState() == ReadyState.OPEN) {
-            try {
-                StreamrClient.this.subs.forEach(StreamrClient.this::resubscribe);
-            } catch (WebsocketNotConnectedException e) {
-                log.error(e);
-                sleepThenReconnect();
-            }
-        } else {
-            sleepThenReconnect();
-        }
-    }
-
     /**
      * Connects the websocket. Blocks until connected, or throws if the connection times out.
      */
     public void connect() throws ConnectionTimeoutException {
-        stayConnected = true;
-        connect(true);
-    }
-
-    private void connect(boolean firstTrial) throws ConnectionTimeoutException {
         if (getState() == ReadyState.OPEN) {
-            log.warn("Trying to connect when already connected to " + options.getWebsocketApiUrl());
             return;
         }
 
-        if (firstTrial) {
-            log.info("Connecting to " + options.getWebsocketApiUrl() + "...");
-            if (websocket == null) {
-                initWebsocket();
+        if (!keepConnected) {
```

Comment:
Synchronize on `lock` _before_ reading `keepConnected`? Now the following execution path is possible:

- Two threads evaluate `!keepConnected` end up inside the `if`
- One thread gets the lock and sets `keepConnected = true` and starts the executorService
- The other thread gets the lock and sets `keepConnected = true` and starts the executorService

Reply:
Good catch. Should have the if-inside the synchronized as well.

Reply:
Fixed